sht3x/sht3x.go

```diff
@@ -0,0 +1,72 @@
+// Package sht3x provides a driver for the SHT3x digital humidity sensor
+// series by Sensirion.
+//
+// Datasheet:
+// https://www.sensirion.com/fileadmin/user_upload/customers/sensirion/Dokumente/0_Datasheets/Humidity/Sensirion_Humidity_Sensors_SHT3x_Datasheet_digital.pdf
+//
+package sht3x
+
+import (
+	"machine"
+	"time"
+)
+
+// Device wraps an I2C connection to a SHT31 device.
+type Device struct {
+	bus     machine.I2C
+	Address uint16
+}
+
+// New creates a new SHT31 connection. The I2C bus must already be
+// configured.
+//
+// This function only creates the Device object, it does not initialize the device.
+// You must call Configure() first in order to use the device itself.
+func New(bus machine.I2C) Device {
+	return Device{
+		bus:     bus,
+		Address: AddressA,
+	}
+}
+
+// Read returns the temperature in celsius milli degrees (ºC/1000).
+func (d *Device) ReadTemperature() (tempMilliCelsius int32, err error) {
+	tempMilliCelsius, _, err = d.ReadTemperatureHumidity()
+	return tempMilliCelsius, err
+}
+
+// Read returns the relative humidity in hundredths of a percent.
+func (d *Device) ReadHumidity() (relativeHumidity int16, err error) {
+	_, relativeHumidity, err = d.ReadTemperatureHumidity()
+	return relativeHumidity, err
+}
+
+// Read returns both the temperature and relative humidity.
+func (d *Device) ReadTemperatureHumidity() (tempMilliCelsius int32, relativeHumidity int16, err error) {
+	var rawTemp, rawHum, errx = d.rawReadings()
+	if errx != nil {
+		err = errx
+		return
+	}
+	tempMilliCelsius = (35000 * int32(rawTemp) / 13107) - 45000
+	relativeHumidity = int16(2000 * int32(rawHum) / 13107)
+	return tempMilliCelsius, relativeHumidity, err
+}
+
+// rawReadings returns the sensor's raw values of the temperature and humidity
+func (d *Device) rawReadings() (uint16, uint16, error) {
+	d.bus.Tx(d.Address, []byte{MEASUREMENT_COMMAND_MSB, MEASUREMENT_COMMAND_LSB}, nil)
+
+	time.Sleep(17 * time.Millisecond)
+
+	data := make([]byte, 5)
```

Comment:
It would be a good idea to allocate this only one time instead of every time that the method is called.

One way to do this is to make `data` a member of the `Device` struct, and initialize it in the `New()` function.

Reply:
Sorry I did not notice this before, but all of the conversation about AVR and limited memory made me think of it. Thanks!